### ISSUE TITLE
Slow document downloads

### ISSUE THREAD

**Issue by anderjf** [benyap/firestore-backfire#183]
Hi benyap,

Firstly, I would like to thank you for version 2.x, it is so better, in the 1.x I was having some problems connecting on the Firebase Emulator, sometimes it connected but sometimes did not, but now it is working fine.

So, now I am finding it is slow to download the documents, when I am using the Firebase Emulator, it downloads about 10 documents and when using the production, it downloads about 40 documents per 5 seconds, I do not know if this slow is because of download document or when saving in the JSON. Do you think there is a way to do it faster? The reason for my question is because my data it big.

It is part of the logs:

> export info    Export data from demo-backup to backfile.ndjson
> export verbose {
>   options: {
>     paths: [ 'xxxx/xxxx/xxxx/xxxx' ],
>     exploreInterval: 10,
>     exploreChunkSize: 5000,
>     downloadChunkSize: 5000,
>     downloadInterval: 10,
>     overwrite: true,
>     debug: true,
>     verbose: true
>   },
>   connection: { project: 'demo-backup', emulator: 'localhost:8180' }
> }
> export debug   Overwriting existing data at backfile.ndjson
> export debug   Progress: 0 exported, 0 exploring, 1 to explore
> export verbose Exploring for subcollections in 1 document
> export verbose Exporting 1 document
> export debug   Progress: 0 exported, 1 exploring, 0 to explore
> export verbose Found 3 subcollections in xxxx/xxxx/xxxx/xxxx
> export verbose Exploring for documents in 3 collections
> export verbose Found 1 document in xxxx/xxxx/xxxx/xxxx/apps
> export verbose Exporting 1 document
> export debug   Progress: 2 exported, 2 exploring, 1 to explore
> export verbose Found 10163 documents in xxxx/xxxx/xxxx/xxxx/xxxx
> export verbose Exporting 5000 documents
> export verbose Found 24 documents in xxxx/xxxx/xxxx/xxxx/xxxx
> export verbose Exploring for subcollections in 5000 documents
> export verbose Exporting 5000 documents
> export debug   Progress: 5002 exported, 5000 exploring, 5188 to explore
> export verbose Found 2 subcollections in xxxx/xxxx/xxxx/xxxx/xxxx/xxxx
> export verbose Exporting 187 documents
> export debug   Progress: 10189 exported, 4990 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4980 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4975 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4966 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4955 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4953 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4942 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4932 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4929 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4918 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4909 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4905 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4894 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4885 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4883 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4873 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4864 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4861 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4851 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4841 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4839 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4831 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4821 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4815 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4809 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4799 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4792 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4788 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4778 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4769 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4766 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4756 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4747 exploring, 5190 to explore
> export debug   Progress: 10189 exported, 4744 exploring, 5190 to explore

**Comment by benyap** [comment]
Hi @anderjf, thanks for using Firestore Backfire and I'm glad it's working better!

A few questions:

1. To clarify, are you saying that it is faster when exporting from the emulator or from a production Firestore instance?
2. I noticed that you've changed some of the advanced config options (`exploreInterval`, `exploreChunkSize`, `downloadInterval`, `downloadChunkSize`). What was the behaviour like with the defaults compared to what you set?
3. When you say that your data is big, do you mean that each individual document is large? Or that you have a lot of records? Any estimates on how big we're talking here? (average document size, number of documents etc.)

---

In my experience, the slowest part of exporting data from Firestore are:
- Listing a large number of documents from a collection using [listDocuments()](https://googleapis.dev/nodejs/firestore/latest/CollectionReference.html#listDocuments)
- Downloading a large number of documents using [getAll(...)](https://googleapis.dev/nodejs/firestore/latest/Firestore.html#getAll)

Unfortunately if you have a collection with a large number of documents, it's just going to be slow given the nature of how [listDocuments is implemented](https://github.com/googleapis/nodejs-firestore/issues/825#issuecomment-569051925). I might try using the [REST API](https://cloud.google.com/firestore/docs/reference/rest/v1/projects.databases.documents/listDocuments) and paginating the calls myself at some point to see if that helps speed things up when I have the time.

Looking at your logs, it doesn't seem like the bottleneck is downloading and saving the JSON data. It managed to export over 10k documents, then it kinda sits around exploring paths without exporting anything.

Assuming that the "Progress" logs are being printed every 5 seconds (which is the default behaviour), I'm a little unsure as to why it's exploring so slowly. It is calling `listDocuments()` for each document it's exploring, but the fact that it hasn't reported it having found any subcollections means that the call itself should have returned very quickly, and it should easily get through all of them within a few seconds. 

If you don't mind, could you share what the structure of your data is? It will be helpful to understand how deeply nested your data is, and how many subcollections/documents there are.

**Comment by anderjf** [comment]
HI @benyap,

Answering your questions:

1. From a production Firestore instance is faster, but not so much.
2. The same results.
3. I was meaning that my database has a lot of records, but the only thing like size, it would be the backup files that I exported from production, which have about 652MB.

Maybe you could also try to use `stream`, I realized that faster than `listDocuments`:

```
  const account = db.collection(path);

  let count = 0;

  account.stream().on('data', (documentSnapshot) => {
    console.log(documentSnapshot.ref.path);
    ++count;
  }).on('end', () => {
    console.log(`Total count is ${count}`);
  });
```

I don't know if it will help, but it is more or less what we have.

```
{
  "col": {
    "doc": {
      "__collections__": {
        "col": {
          "doc": {
            "__collections__": {
              "col": {
                "doc": {
                  "__collections__": {
                    "col": {  /** in this collection we have about 11k documents **/
                      "doc": {
                        "field": "value",
                        "field": "value"
                      },
                      "doc": {
                        "field": "value",
                        "field": "value"
                      },
                      "doc": {
                        "field": "value",
                        "field": "value"
                      }
                    },
                    "col": { /** in this collection we have about 200k documents **/
                      "doc": {
                        "field": "value",
                        "field": "value"
                      },
                      "doc": {
                        "field": "value",
                        "field": "value"
                      },
                      "doc": {
                        "field": "value",
                        "field": "value"
                      }
                    }
                  },
                  "field": "value",
                  "field": "value"
                }
              }
            },
            "field": "value",
            "field": "value"
          }
        }
      }
    }
  }
}
```

Thank you for your help.

**Comment by benyap** [comment]
Hey @anderjf,

I did some digging and tried using `stream()` as you suggested, and it indeed was faster as it allows concurrent processing of data as it streams in. This change has been released in [v2.1.0](https://github.com/benyap/firestore-backfire/releases/tag/v2.1.0), so please give it a shot and let me know how it goes!

Closing this for now. Feel free to reopen if you run into any issues related to this.

**Comment by anderjf** [comment]
Hi @benyap,

I'm glad to have helped.

So, I tested it, and it is working well and downloading faster, but the bottleneck now is in exploring if the document has any collection, see the logs.

```
export 8:13:32 am info    Export data from demo-backup to backfile.ndjson
export 8:13:32 am verbose {
  options: {
    paths: [ 'xxxx/xxxx/xxxx/xxxx/risksxxxx' ],
    overwrite: true,
    debug: true,
    verbose: true
  },
  connection: { project: 'demo-backup', emulator: 'localhost:8180' }
}
export 8:13:32 am debug   Overwriting existing data at backfile.ndjson
export 8:13:32 am debug   Progress: 0 exported, 1 to export, 0 exploring, 1 to explore, 0 reads
export 8:13:32 am debug   Exploring for documents in 1 collection
export 8:13:33 am verbose Exporting 9736 documents...
export 8:13:33 am debug   Successfully exported 9736 documents
export 8:13:33 am debug   Exploring for subcollections in 1000 documents
export 8:13:34 am verbose Exporting 427 documents...
export 8:13:34 am debug   Successfully exported 427 documents
export 8:13:37 am debug   Progress: 10163 exported, 0 to export, 991 exploring, 9163 to explore, 10163 reads
export 8:13:42 am debug   Progress: 10163 exported, 0 to export, 979 exploring, 9163 to explore, 10163 reads
export 8:13:47 am debug   Progress: 10163 exported, 0 to export, 966 exploring, 9163 to explore, 10163 reads
export 8:13:52 am debug   Progress: 10163 exported, 0 to export, 954 exploring, 9163 to explore, 10163 reads
export 8:13:57 am debug   Progress: 10163 exported, 0 to export, 942 exploring, 9163 to explore, 10163 reads
export 8:14:02 am debug   Progress: 10163 exported, 0 to export, 930 exploring, 9163 to explore, 10163 reads
export 8:14:07 am debug   Progress: 10163 exported, 0 to export, 917 exploring, 9163 to explore, 10163 reads
export 8:14:12 am debug   Progress: 10163 exported, 0 to export, 905 exploring, 9163 to explore, 10163 reads
export 8:14:17 am debug   Progress: 10163 exported, 0 to export, 893 exploring, 9163 to explore, 10163 reads
export 8:14:22 am debug   Progress: 10163 exported, 0 to export, 881 exploring, 9163 to explore, 10163 reads
export 8:14:27 am debug   Progress: 10163 exported, 0 to export, 869 exploring, 9163 to explore, 10163 reads
export 8:14:32 am debug   Progress: 10163 exported, 0 to export, 857 exploring, 9163 to explore, 10163 reads
export 8:14:37 am debug   Progress: 10163 exported, 0 to export, 845 exploring, 9163 to explore, 10163 reads
export 8:14:42 am debug   Progress: 10163 exported, 0 to export, 832 exploring, 9163 to explore, 10163 reads
export 8:14:47 am debug   Progress: 10163 exported, 0 to export, 820 exploring, 9163 to explore, 10163 reads
export 8:14:52 am debug   Progress: 10163 exported, 0 to export, 808 exploring, 9163 to explore, 10163 reads
export 8:14:57 am debug   Progress: 10163 exported, 0 to export, 796 exploring, 9163 to explore, 10163 reads
export 8:15:02 am debug   Progress: 10163 exported, 0 to export, 784 exploring, 9163 to explore, 10163 reads
```

In this case, there are 10163 documents, and it will check for each document if there are some subcollection, I don't know if has some way to do that faster, maybe to change this verification to be a parallel verification, using a Promise.

**Comment by benyap** [comment]
Thanks for the feedback @anderjf. Will look into this soon 😊

**Comment by benyap** [comment]
Hi @anderjf, an enhancement to improve the subcollection exploration speed has been released in [v2.2.0](https://github.com/benyap/firestore-backfire/releases/tag/v2.2.0) 😊
